File: display_test.go
package main

import (
	"bytes"
	"io"
	"os"
	"strings"
	"testing"
	"time"
)

func captureOutput(f func()) string {
	old := os.Stdout
	r, w, _ := os.Pipe()
	os.Stdout = w

	f()

	w.Close()
	os.Stdout = old

	var buf bytes.Buffer
	io.Copy(&buf, r)
	return buf.String()
}

func TestDisplayStatsNfsiostat(t *testing.T) {
	mount := &NFSMount{
		Device:     "server:/export",
		MountPoint: "/mnt/nfs",
		Events: &NFSEvents{
			VFSOpen:         100,
			InodeRevalidate: 200,
			DataInvalidate:  50,
			AttrInvalidate:  25,
		},
	}

	stats := []*DeltaStats{
		{
			Operation:    "READ",
			DeltaOps:     100,
			DeltaSent:    1024,
			DeltaRecv:    2048,
			DeltaBytes:   3072,
			DeltaRTT:     500,
			DeltaExec:    1000,
			DeltaQueue:   100,
			DeltaErrors:  2,
			DeltaRetrans: 1,
			IOPS:         100.0,
			AvgRTT:       5.0,
			AvgExec:      10.0,
			AvgQueue:     1.0,
			KBPerOp:      3.0,
			KBPerSec:     300.0,
		},
		{
			Operation:    "WRITE",
			DeltaOps:     50,
			DeltaSent:    2048,
			DeltaRecv:    512,
			DeltaBytes:   2560,
			DeltaRTT:     250,
			DeltaExec:    500,
			DeltaQueue:   50,
			DeltaErrors:  0,
			DeltaRetrans: 0,
			IOPS:         50.0,
			AvgRTT:       5.0,
			AvgExec:      10.0,
			AvgQueue:     1.0,
			KBPerOp:      2.5,
			KBPerSec:     125.0,
		},
	}

	previousMount := &NFSMount{
		Events: &NFSEvents{
			VFSOpen:         90,
			InodeRevalidate: 180,
			DataInvalidate:  45,
			AttrInvalidate:  20,
		},
	}

	tests := []struct {
		name     string
		showAttr bool
		verify   func(t *testing.T, output string)
	}{
		{
			name:     "without attributes",
			showAttr: false,
			verify: func(t *testing.T, output string) {
				if !strings.Contains(output, "server:/export mounted on /mnt/nfs:") {
					t.Error("Missing mount header")
				}
				if !strings.Contains(output, "ops/s") {
					t.Error("Missing ops/s header")
				}
				if !strings.Contains(output, "read:") {
					t.Error("Missing READ operation")
				}
				if !strings.Contains(output, "write:") {
					t.Error("Missing WRITE operation")
				}
				if strings.Contains(output, "VFS opens") {
					t.Error("Should not show VFS opens when showAttr is false")
				}
			},
		},
		{
			name:     "with attributes",
			showAttr: true,
			verify: func(t *testing.T, output string) {
				if !strings.Contains(output, "10 VFS opens") {
					t.Error("Missing VFS opens count")
				}
				if !strings.Contains(output, "20 inoderevalidates") {
					t.Error("Missing inode revalidates count")
				}
				if !strings.Contains(output, "5 page cache invalidations") {
					t.Error("Missing page cache invalidations")
				}
				if !strings.Contains(output, "5 attribute cache invalidations") {
					t.Error("Missing attribute cache invalidations")
				}
			},
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			output := captureOutput(func() {
				displayStatsNfsiostat(mount, stats, previousMount, tt.showAttr)
			})
			tt.verify(t, output)
		})
	}
}

func TestDisplayStatsSimple(t *testing.T) {
	mount := &NFSMount{
		Device:     "server:/export",
		MountPoint: "/mnt/nfs",
	}

	stats := []*DeltaStats{
		{
			Operation: "READ",
			DeltaOps:  100,
			IOPS:      100.0,
			AvgRTT:    5.123,
			AvgExec:   10.456,
			KBPerSec:  1024.0,
			KBPerOp:   10.24,
		},
		{
			Operation: "WRITE",
			DeltaOps:  50,
			IOPS:      50.0,
			AvgRTT:    6.789,
			AvgExec:   12.345,
			KBPerSec:  512.0,
			KBPerOp:   10.24,
		},
	}

	timestamp := time.Date(2024, 1, 1, 12, 0, 0, 0, time.UTC)

	tests := []struct {
		name          string
		showBandwidth bool
		verify        func(t *testing.T, output string)
	}{
		{
			name:          "without bandwidth",
			showBandwidth: false,
			verify: func(t *testing.T, output string) {
				if !strings.Contains(output, "server:/export mounted on 01/01/2024 12:00:00 PM") {
					t.Error("Missing mount header with timestamp")
				}
				if !strings.Contains(output, "Operation") && !strings.Contains(output, "IOPS") {
					t.Error("Missing column headers")
				}
				if !strings.Contains(output, "READ") {
					t.Error("Missing READ operation")
				}
				if !strings.Contains(output, "WRITE") {
					t.Error("Missing WRITE operation")
				}
				if strings.Contains(output, "MB/s") || strings.Contains(output, "KB/op") {
					t.Error("Should not show bandwidth columns when showBandwidth is false")
				}
			},
		},
		{
			name:          "with bandwidth",
			showBandwidth: true,
			verify: func(t *testing.T, output string) {
				if !strings.Contains(output, "MB/s") {
					t.Error("Missing MB/s column")
				}
				if !strings.Contains(output, "KB/op") {
					t.Error("Missing KB/op column")
				}
			},
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			output := captureOutput(func() {
				displayStatsSimple(mount, stats, tt.showBandwidth, timestamp)
			})
			tt.verify(t, output)
		})
	}
}

func TestDisplayStatsSimpleEmptyStats(t *testing.T) {
	mount := &NFSMount{
		Device:     "server:/export",
		MountPoint: "/mnt/nfs",
	}

	// Test with empty stats
	output := captureOutput(func() {
		displayStatsSimple(mount, []*DeltaStats{}, false, time.Now())
	})

	if output != "" {
		t.Error("Expected no output for empty stats")
	}

	// Test with nil stats
	output = captureOutput(func() {
		displayStatsSimple(mount, nil, false, time.Now())
	})

	if output != "" {
		t.Error("Expected no output for nil stats")
	}
}

func TestDisplayStatsNfsiostatWithNilStats(t *testing.T) {
	mount := &NFSMount{
		Device:     "server:/export",
		MountPoint: "/mnt/nfs",
		Events:     &NFSEvents{},
	}

	// Test with stats containing nil entries
	stats := []*DeltaStats{
		nil,
		{
			Operation: "READ",
			DeltaOps:  0, // Zero ops should be skipped
		},
		{
			Operation: "WRITE",
			DeltaOps:  10,
			IOPS:      10.0,
			AvgRTT:    5.0,
			AvgExec:   10.0,
		},
	}

	output := captureOutput(func() {
		displayStatsNfsiostat(mount, stats, nil, false)
	})

	if !strings.Contains(output, "write:") {
		t.Error("Should display WRITE operation")
	}
	if strings.Contains(output, "read:") {
		t.Error("Should not display READ operation with 0 ops")
	}
}

File: tests/display_test.rs
use nfs_gaze::{DeltaStats, NFSMount};
use nfs_gaze::display::display_stats_simple;
use std::collections::HashMap;
use chrono::{Utc, TimeZone};
use std::io::{self, Write};

// Mock writer to capture output instead of writing to stdout
struct MockWriter {
    pub buffer: Vec<u8>,
}

impl MockWriter {
    fn new() -> Self {
        Self { buffer: Vec::new() }
    }

    fn to_string(&self) -> String {
        String::from_utf8(self.buffer.clone()).unwrap()
    }
}

impl Write for MockWriter {
    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
        self.buffer.extend_from_slice(buf);
        Ok(buf.len())
    }

    fn flush(&mut self) -> io::Result<()> {
        Ok(())
    }
}

#[test]
fn test_display_stats_simple_without_bandwidth() {
    let mount = NFSMount {
        device: "server:/export".to_string(),
        mount_point: "/mnt/nfs".to_string(),
        server: "server".to_string(),
        export: "/export".to_string(),
        age: 0,
        operations: HashMap::new(),
        events: None,
        bytes_read: 0,
        bytes_write: 0,
    };

    let stats = vec![
        DeltaStats {
            operation: "READ".to_string(),
            delta_ops: 100,
            delta_bytes: 0,
            delta_sent: 0,
            delta_recv: 0,
            delta_rtt: 0,
            delta_exec: 0,
            delta_queue: 0,
            delta_errors: 0,
            delta_retrans: 0,
            avg_rtt: 1.5,
            avg_exec: 2.0,
            avg_queue: 0.0,
            kb_per_op: 10.24,
            kb_per_sec: 1024.0,
            iops: 100.0,
        },
        DeltaStats {
            operation: "WRITE".to_string(),
            delta_ops: 50,
            delta_bytes: 0,
            delta_sent: 0,
            delta_recv: 0,
            delta_rtt: 0,
            delta_exec: 0,
            delta_queue: 0,
            delta_errors: 0,
            delta_retrans: 0,
            avg_rtt: 2.5,
            avg_exec: 3.0,
            avg_queue: 0.0,
            kb_per_op: 10.24,
            kb_per_sec: 512.0,
            iops: 50.0,
        },
    ];

    let timestamp = Utc.with_ymd_and_hms(2024, 1, 1, 12, 0, 0).unwrap();
    let mut writer = MockWriter::new();

    display_stats_simple(&mut writer, &mount, &stats, false, &timestamp).unwrap();
    let output = writer.to_string();

    assert!(output.contains("READ"), "Output should contain READ operation");
    assert!(output.contains("WRITE"), "Output should contain WRITE operation");
    assert!(output.contains("100.0"), "Output should contain IOPS for READ");
    assert!(!output.contains("MB/s"), "Output should not contain bandwidth columns");
}

#[test]
fn test_display_stats_simple_with_bandwidth() {
    let mount = NFSMount {
        device: "server:/export".to_string(),
        mount_point: "/mnt/nfs".to_string(),
        server: "server".to_string(),
        export: "/export".to_string(),
        age: 0,
        operations: HashMap::new(),
        events: None,
        bytes_read: 0,
        bytes_write: 0,
    };

    let stats = vec![
        DeltaStats {
            operation: "READ".to_string(),
            delta_ops: 100,
            delta_bytes: 0,
            delta_sent: 0,
            delta_recv: 0,
            delta_rtt: 0,
            delta_exec: 0,
            delta_queue: 0,
            delta_errors: 0,
            delta_retrans: 0,
            avg_rtt: 1.5,
            avg_exec: 2.0,
            avg_queue: 0.0,
            kb_per_op: 10.24,
            kb_per_sec: 1024.0,
            iops: 100.0,
        },
        DeltaStats {
            operation: "WRITE".to_string(),
            delta_ops: 50,
            delta_bytes: 0,
            delta_sent: 0,
            delta_recv: 0,
            delta_rtt: 0,
            delta_exec: 0,
            delta_queue: 0,
            delta_errors: 0,
            delta_retrans: 0,
            avg_rtt: 2.5,
            avg_exec: 3.0,
            avg_queue: 0.0,
            kb_per_op: 10.24,
            kb_per_sec: 512.0,
            iops: 50.0,
        },
    ];

    let timestamp = Utc.with_ymd_and_hms(2024, 1, 1, 12, 0, 0).unwrap();
    let mut writer = MockWriter::new();

    display_stats_simple(&mut writer, &mount, &stats, true, &timestamp).unwrap();
    let output = writer.to_string();

    assert!(output.contains("READ"), "Output should contain READ operation");
    assert!(output.contains("WRITE"), "Output should contain WRITE operation");
    assert!(output.contains("MB/s"), "Output should contain bandwidth column");
    assert!(output.contains("KB/op"), "Output should contain KB/op column");
}

#[test]
fn test_display_stats_simple_empty_stats() {
    let mount = NFSMount {
        device: "server:/export".to_string(),
        mount_point: "/mnt/nfs".to_string(),
        server: "server".to_string(),
        export: "/export".to_string(),
        age: 0,
        operations: HashMap::new(),
        events: None,
        bytes_read: 0,
        bytes_write: 0,
    };

    let stats: Vec<DeltaStats> = vec![];
    let timestamp = Utc.with_ymd_and_hms(2024, 1, 1, 12, 0, 0).unwrap();
    let mut writer = MockWriter::new();

    display_stats_simple(&mut writer, &mount, &stats, false, &timestamp).unwrap();
    let output = writer.to_string();

    // Should have minimal output for empty stats
    assert!(output.is_empty() || output.trim().is_empty(), "Output should be empty for empty stats");
}

#[test]
fn test_format_duration() {
    use nfs_gaze::display::format_duration;

    assert_eq!(format_duration(500), "0.5ms");
    assert_eq!(format_duration(1500), "1.5ms");
    assert_eq!(format_duration(10000), "10.0ms");
    assert_eq!(format_duration(0), "0.0ms");
}

#[test]
fn test_format_rate() {
    use nfs_gaze::display::format_rate;

    assert_eq!(format_rate(0.0), "0.0");
    assert_eq!(format_rate(1.5), "1.5");
    assert_eq!(format_rate(100.0), "100.0");
    assert_eq!(format_rate(1000.5), "1000.5");
}

#[test]
fn test_format_bandwidth() {
    use nfs_gaze::display::format_bandwidth;

    // Test KB/s
    assert_eq!(format_bandwidth(512.0), "0.5");
    assert_eq!(format_bandwidth(1024.0), "1.0");
    assert_eq!(format_bandwidth(1536.0), "1.5");

    // Test zero
    assert_eq!(format_bandwidth(0.0), "0.0");

    // Test large values
    assert_eq!(format_bandwidth(10240.0), "10.0");
}